qt/scientific_interfaces/Indirect/IndirectFitDataTableModel.cpp

```diff
@@ -0,0 +1,349 @@
+//   Institut Laue - Langevin & CSNS, Institute of High Energy Physics, CAS
+// SPDX - License - Identifier: GPL - 3.0 +
+#include <string>
+
+#include "DllConfig.h"
+#include "IndirectFitData.h"
+#include "IndirectFitDataTableModel.h"
+#include "MantidAPI/AnalysisDataService.h"
+#include "MantidAPI/MatrixWorkspace.h"
+
+using namespace MantidQt::MantidWidgets;
+
+namespace {
+
+bool equivalentWorkspaces(const Mantid::API::MatrixWorkspace_const_sptr &lhs,
+                          const Mantid::API::MatrixWorkspace_const_sptr &rhs) {
+  if (!lhs || !rhs)
+    return false;
+  else if (lhs->getName() == "" && rhs->getName() == "")
+    return lhs == rhs;
+  return lhs->getName() == rhs->getName();
+}
+
+bool doesExistInADS(std::string const &workspaceName) {
+  return Mantid::API::AnalysisDataService::Instance().doesExist(workspaceName);
+}
+} // namespace
+
+namespace MantidQt {
+namespace CustomInterfaces {
+namespace IDA {
+
+IndirectFitDataTableModel::IndirectFitDataTableModel()
+    : m_fittingDataSingle(std::make_unique<std::vector<IndirectFitData>>()),
+      m_resolutionsSingle(std::make_unique<std::vector<std::weak_ptr<Mantid::API::MatrixWorkspace>>>()),
+      m_fittingDataMultiple(std::make_unique<std::vector<IndirectFitData>>()),
+      m_resolutionsMultiple(std::make_unique<std::vector<std::weak_ptr<Mantid::API::MatrixWorkspace>>>()) {
+
+  m_fittingData = m_fittingDataSingle.get();
+  m_resolutions = m_resolutionsSingle.get();
+}
+
+bool IndirectFitDataTableModel::hasWorkspace(std::string const &workspaceName) const {
+  auto const names = getWorkspaceNames();
+  auto const iter = std::find(names.begin(), names.end(), workspaceName);
+  return iter != names.end();
+}
+
+Mantid::API::MatrixWorkspace_sptr IndirectFitDataTableModel::getWorkspace(TableDatasetIndex index) const {
+  if (index < m_fittingData->size())
+    return m_fittingData->at(index.value).workspace();
+  return nullptr;
+}
+
+FunctionModelSpectra IndirectFitDataTableModel::getSpectra(TableDatasetIndex index) const {
+  if (index < m_fittingData->size())
+    return m_fittingData->at(index.value).spectra();
+  return FunctionModelSpectra("");
+}
+
+TableDatasetIndex IndirectFitDataTableModel::getNumberOfWorkspaces() const {
+  return TableDatasetIndex{m_fittingData->size()};
+}
+
+bool IndirectFitDataTableModel::isMultiFit() const { return getNumberOfWorkspaces().value > 1; }
+
+size_t IndirectFitDataTableModel::getNumberOfSpectra(TableDatasetIndex index) const {
+  if (index < m_fittingData->size())
+    return m_fittingData->at(index.value).numberOfSpectra().value;
+  else
+    throw std::runtime_error("Cannot find the number of spectra for a workspace: the workspace "
+                             "index provided is too large.");
+  return 1;
+}
+
+size_t IndirectFitDataTableModel::getNumberOfDomains() const {
+  size_t sum{0};
+  for (auto fittingData : *m_fittingData) {
+    sum += fittingData.numberOfSpectra().value;
+  }
+  return sum;
```

Comment:
Would it be possible to use `std::accumulate` for this by pass in a lambda?

Reply:
I don't think so, since the itterator is over a vector of IndirectFitData and we want size_t out i get an error trying to convert with it.

Reply:
Would something like this not work?
```suggestion
  std::size_t init(0);
  auto const getNumberOfSpectra = [](auto const &fittingData){ return fittingData.numberOfSpectra().value; }
  return std::accumulate (m_fittingData.cbegin(), m_fittingData.cend(), init, getNumberOfSpectra);
```

Reply:
I figured out how to do it right. thanks